File: Checker.cpp
#include <bits/stdc++.h>
using namespace std;
const int N = 8;
const int n_players = 2;
char marks[n_players] = {'w', 'b'};
struct marks_player
{
    bool isKing = false;
    char mark = ' ';
};
marks_player grid[N][N];
void clear_grid()
{
    for(int i=0;i < N; i++)
    {
        for(int j=0;j < N; j++)
            grid[i][j].mark = ' ';
    }
}
void set_elements()
{
    marks_player player;
    player.mark = 'w';
    grid[0][0] = player; grid[0][2] = player; grid[0][4] = player; grid[0][6] = player;
    grid[1][1] = player; grid[1][3] = player; grid[1][5] = player; grid[1][7] = player;
    grid[2][0] = player; grid[2][2] = player; grid[2][4] = player; grid[2][6] = player;
    player.mark = 'b';
    grid[5][1] = player; grid[5][3] = player; grid[5][5] = player; grid[5][7] = player;
    grid[6][0] = player; grid[6][2] = player; grid[6][4] = player; grid[6][6] = player;
    grid[7][1] = player; grid[7][3] = player; grid[7][5] = player; grid[7][7] = player;
}
void print_grid()
{
 for(int i=0;i<8;i++)
 {
  cout<<"  --------------------------------"<<endl;
  for(int j = 0; j < N; j++)
  {
      if(grid[i][j].isKing)
        cout<<" | " << char(int(grid[i][j].mark) - 32);
      else
        cout<<" | " << grid[i][j].mark;
  }
  cout << " | " << endl;
 }
 cout<<"  --------------------------------"<<endl;
}
bool valid_position(int i, int j, int i2, int j2, char mark)
{
    if(grid[i][j].mark != mark || grid[i2][j2].mark != ' ')
        return false;
    if((i < 0 || i > N ) || (j < 0 || j > N) || (i2 < 0 || i2 > N ) || (j2 < 0 || j2 > N))
        return false;
    return true;
}
void move_mark(int i, int j, int i2, int j2)
{
    grid[i2][j2] = grid[i][j];
    grid[i][j].mark = ' ';
    grid[i][j].isKing = false;
    if (grid[i2][j2].mark == 'w' && i2 == N-1)
        grid[i2][j2].isKing = true;
    if(grid[i2][j2].mark == 'b' && i2 == 0)
        grid[i2][j2].isKing = true;
}
bool out (int i, int j, int i2, int j2)
{
    if(!grid[i][j].isKing)
    {
        if (grid[i][j].mark == 'b' && grid[i-1][j+1].mark == 'w' && i2 == i - 2 && j2 == j + 2)
        {
            grid[i-1][j+1].mark = ' ';
            return true;
        }
        else if (grid[i][j].mark == 'b' && grid[i-1][j-1].mark == 'w' && i2 == i - 2 && j2 == j - 2)
        {
            grid[i-1][j-1].mark = ' ';
            return true;
        }
        else if (grid[i][j].mark == 'w' && grid[i+1][j-1].mark == 'b' && i2 == i + 2 && j2 == j - 2)
        {
            grid[i+1][j-1].mark = ' ';
            return true;
        }
        else if (grid[i][j].mark == 'w' && grid[i+1][j+1].mark == 'b' && i2 == i + 2 && j2 == j + 2)
        {
            grid[i+1][j+1].mark = ' ';
            return true;
        }
        return false;
    }
    else
    {
        if (grid[i][j].mark != grid[i-1][j+1].mark && grid[i-1][j+1].mark != ' ' && grid[i][j].mark != ' '  && i2 == i - 2 && j2 == j + 2)
        {
            grid[i-1][j+1].mark = ' ';
            return true;
        }
        else if (grid[i][j].mark != grid[i-1][j-1].mark && grid[i-1][j-1].mark != ' ' && grid[i][j].mark != ' ' && i2 == i - 2 && j2 == j - 2)
        {
            grid[i-1][j-1].mark = ' ';
            return true;
        }
        else if (grid[i][j].mark != grid[i+1][j-1].mark && grid[i+1][j-1].mark != ' ' && grid[i][j].mark != ' ' &&  i2 == i + 2 && j2 == j - 2)
        {
            grid[i+1][j-1].mark = ' ';
            return true;
        }
        else if (grid[i][j].mark != grid[i+1][j+1].mark && grid[i+1][j+1].mark != ' ' && grid[i][j].mark != ' ' &&  i2 == i + 2 && j2 == j + 2)
        {
            grid[i+1][j+1].mark = ' ';
            return true;
        }
        return false;
    }
}
bool valid_move(int i, int j, int i2, int j2)
{
    if(out(i,j,i2,j2))
        return true;
    else
    {
        if((grid[i][j].mark == 'w') && (i2 <= i || i2 != i + 1) && (!grid[i][j].isKing))
                return false;
        if((grid[i][j].mark == 'w') && ((j2 != j + 1) && (j2 != j - 1)) && (!grid[i][j].isKing))
                return false;
        if((grid[i][j].mark == 'b') && (i2 >= i || i2 != i - 1) && (!grid[i][j].isKing))
                return false;
        if((grid[i][j].mark == 'b') && ((j2 != j + 1) && (j2 != j - 1))&& (!grid[i][j].isKing))
                return false;
         if((grid[i][j].isKing) && (abs(i - i2) != abs(j - j2)) || (abs(i - i2) == abs(j - j2) && abs(i-i2) != 1))
                return false;
    }
    return true;
}
bool win()
{
    int w = 0, b = 0;
    for(int k = 0; k  < N; k++)
    {
        for(int l = 0; l < N; l++)
        {
            if(grid[k][l].mark == 'w')
                w++;
            if(grid[k][l].mark == 'b')
                b++;
        }
    }
    if(b == 0)
    {
        cout<<"W win !!"<<endl;
        return true;
    }
    else if(w == 0)
    {
        cout<<"B win !!"<<endl;
        return true;
    }
    else
        return false;
}
int read_input(int &i , int &j, int &i2, int &j2,char mark)
{
    cout<<"Enter cells indices: "<<endl;
    cin>>i>>j>>i2>>j2;
    while(!valid_position(i,j,i2,j2,mark) || !valid_move(i,j,i2,j2))
    {
        cout<<"Enter valid cells indices :\ "<<endl;
         cin>>i>>j>>i2>>j2;
    }
}
void play()
{
        cout<<"Checker Game..."<<endl;
        cout<<"-------------------------"<<endl;
        int player = 0;
        while (true)
        {
            print_grid();
            cout<<"Player "<<marks[player]<<" is playing now"<<endl;
            int i,j,i2,j2;
            read_input(i,j,i2,j2,marks[player]);
            move_mark(i,j,i2,j2);
            if(win())
              {
                 print_grid();
                 break;
              }
            player = (player + 1) % n_players;
        }
}
int main()
{
    while(true)
    {
        clear_grid();
        set_elements();
        play();
        char c;
        cout<<"Play Again [Y/N]";
        cin>>c;
        if(c != 'y' && c!= 'Y')
            break;
    }
}

File: Checker.py
N = 8
M = 8
n_players = 2
marks = ['w', 'b']
grid = [[' ']*(M) for _ in range(N)]
def clear_grid():
    for i in range(N):
        for j in range(M):
            grid[i][j] = ' '

def set_elements():
    mark = 'w'
    grid[0][0], grid[0][2], grid[0][4], grid[0][6] = mark,mark,mark,mark
    grid[1][1], grid[1][3], grid[1][5], grid[1][7] = mark,mark,mark,mark
    grid[2][0], grid[2][2], grid[2][4], grid[2][6] = mark,mark,mark,mark
    mark = 'b'
    grid[5][1] = mark
    grid[5][3], grid[5][5], grid[5][7] = mark,mark,mark
    grid[6][0], grid[6][2], grid[6][4], grid[6][6] = mark,mark,mark,mark
    grid[7][1], grid[7][3], grid[7][5], grid[7][7] = mark,mark,mark,mark

def print_grid():
    for i in range(N):
        print("  -----------------------------------------------")
        for j in range(M):
            if grid[i][j] == 'B' or grid[i][j] == 'W':
                print(" | ", chr(ord(grid[i][j]) - 32), end = ' ')
            else:
                print(" | ", grid[i][j], end = ' ')
        print(" | ")
    print("  -----------------------------------------------")

def valid_position( i,  j,  i2,  j2,  mark):
    if not grid[i][j] == mark or not grid[i2][j2] == ' ':
        return False
    if (i < 0 or i > N ) or (j < 0 or j > N) or (i2 < 0 or i2 > N) or (j2 < 0 or j2 > N):
        return False
    return True

def move_mark( i,  j,  i2,  j2):
    grid[i2][j2] = grid[i][j]
    grid[i][j] = ' '
    if grid[i2][j2] == 'w' and i2 == N-1:
        grid[i2][j2] = 'W'
    if grid[i2][j2] == 'b' and i2 == 0:
        grid[i2][j2] = 'B'

def out( i, j, i2, j2):
    if not grid[i][j] == 'W' or not grid[i][j] == 'B':
        if grid[i][j] == 'b' and grid[i-1][j+1] == 'w' and i2 == i - 2 and j2 == j + 2:
            grid[i-1][j+1] = ' '
            return True
        elif grid[i][j] == 'b' and grid[i-1][j-1] == 'w' and i2 == i - 2 and j2 == j - 2:
            grid[i-1][j-1] = ' '
            return True
        elif grid[i][j] == 'w' and grid[i+1][j-1] == 'b' and i2 == i + 2 and j2 == j - 2:
            grid[i+1][j-1] = ' '
            return True
        elif grid[i][j] == 'w' and grid[i+1][j+1] == 'b' and i2 == i + 2 and j2 == j + 2:
            grid[i+1][j+1] = ' '
            return True
        return False
    else:
        if not grid[i][j] == grid[i-1][j+1] and not grid[i-1][j+1] == ' ' and not grid[i][j] == ' ' and i2 == i - 2 and j2 == j + 2:
            grid[i-1][j+1] = ' '
            return True
        elif not grid[i][j] == grid[i-1][j-1] and not grid[i-1][j-1] == ' ' and not grid[i][j] == ' ' and i2 == i - 2 and j2 == j - 2:
            grid[i-1][j-1] = ' '
            return True
        elif not grid[i][j] == grid[i+1][j-1] and not grid[i+1][j-1] == ' ' and not grid[i][j] == ' ' and  i2 == i + 2 and j2 == j - 2:
            grid[i+1][j-1] = ' '
            return True
        elif not grid[i][j] == grid[i+1][j+1] and not grid[i+1][j+1] == ' ' and not grid[i][j] == ' ' and  i2 == i + 2 and j2 == j + 2:
            grid[i+1][j+1] = ' '
            return True
        return False

def valid_move(i, j, i2, j2):
    if out(i,j,i2,j2):
        return True
    else:
        if grid[i][j] == 'w' and (i2 <= i or not i2 == i + 1):
            return False
        if grid[i][j] == 'w' and not ((j2 == j + 1) and not (j2 == j - 1)):
            return False
        if grid[i][j] == 'b' and (i2 >= i or not i2 == i - 1):
            return False
        if grid[i][j] == 'b' and not ((j2 == j + 1) and not (j2 == j - 1)):
            return False
        if (grid[i][j] == 'W' or grid[i][j] == 'B') and not (abs(i - i2) == abs(j - j2)) or (abs(i - i2) == abs(j - j2) and not abs(i-i2) == 1):
            return False
    return True
def win():
    w = 0
    b = 0
    for k in range(N):
        for l in range(M):
            if grid[k][l] == 'w' or grid[k][l] == 'W':
                w += 1
            if grid[k][l]  == 'b' or  grid[k][l] == 'B':
                b += 1
    if b == 0 :
        print("W win !!")
        return True
    elif w == 0:
        print("B win !!")
        return True
    else:
        return False
def read_input(i , j, i2, j2, mark):
    i, j, i2, j2 = map(int, input("Enter cells indices: ").split())
    while not valid_position(i,j,i2,j2,mark) or not valid_move(i,j,i2,j2):
        i, j, i2, j2 = map(int, input("Enter valid cells indices : ").split())
    return i,j,i2,j2

def play():
    print("Checker Game...")
    print("-------------------------")
    player = 0
    while (True):
        print_grid()
        print("Player ", marks[player], " is playing now")
        i,j,i2,j2 = 0,0,0,0
        i,j,i2,j2 = read_input(i, j, i2, j2, marks[player])
        move_mark(i,j,i2,j2)
        if win():
            print_grid()
            break
        player = (player + 1) % n_players

while True:
    clear_grid()
    set_elements()
    play()
    c = input("Play Again [Y/N]")
    if not c == 'y' and not c == 'Y':
        break
